From a48b57a14bba42001f7ca674e4958ab5395c6e5e Mon Sep 17 00:00:00 2001
From: Miles Hu <milehu@codeaurora.org>
Date: Fri, 4 Oct 2019 15:49:52 -0700
Subject: [PATCH] ath11k: support hostapd config to enable/disable he mu-mimo
 add he cap cross-check in beacon_enable event to enable he mu-mimo in ap
 mode. add recalc in bss_assoc to enable he mu-mimo in sta mode.

Signed-off-by: Miles Hu <milehu@codeaurora.org>
---
 include/net/mac80211.h                |   1 +
 net/mac80211/cfg.c                    |   5 +-
 4 files changed, 138 insertions(+), 11 deletions(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -708,6 +708,7 @@ struct ieee80211_bss_conf {
 		u16 nss_set;
 	} he_oper;
 	struct ieee80211_he_obss_pd he_obss_pd;
+	struct ieee80211_he_cap_elem  he_cap_elem;
 	struct cfg80211_he_bss_color he_bss_color;
 	struct ieee80211_fils_discovery fils_discovery;
 	u32 unsol_bcast_probe_resp_interval;
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -1202,6 +1202,11 @@ static int ieee80211_start_ap(struct wip
 			return err;
 	}
 
+	if (params->he_cap) {
+		memcpy(&sdata->vif.bss_conf.he_cap_elem, params->he_cap,
+		       sizeof(*params->he_cap));
+	}
+
 	mutex_lock(&local->mtx);
 	err = ieee80211_link_use_channel(link, &params->chandef,
 					 IEEE80211_CHANCTX_SHARED);
